### PR TITLE
Fix harvesters losing their last harvesting position when carried by carryall.

### LINKED ISSUE #16775: D2K: Automatic carryalls override player harvest positions.
If a carryall picks up the harvester, the harvester will always be delivered to the closest spice patch.

Should be a  `new FindAndDeliverResources(self, self.World.Map.CellContaining(self.CenterPosition))` at https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Activities/DeliverResources.cs#L61 and a `new FindAndDeliverResources(self, targetCell)` at https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Activities/HarvestResource.cs#L59 probably. Fix untested.

(The original reason I noticed this is because chrono miners also use these interfaces and those broke down too, which I fixed within the trait code instead.)

### PR DESCRIPTION
**Pull request by tovl**
Fixes #16775 

This is just a quick fix for next release. A proper fix would be the same as https://github.com/OpenRA/OpenRA/pull/16721#issuecomment-511197835 which would also conserve activities queued afterwards.

### CODE REVIEW

**Comment by GraionDilach** [comment]
Oh, so I was correct with my proposals? Good to know.

**Comment by pchote** [comment]
Testing this shows another fairly bad regression with carryalls:
1. Have a harvester and carryall working in normal automatic mode
2. Wait for the harvester to request a pickup, then quickly issue a harvest order to a different field

Things get weird, depending on the exact timing. The most common case is that the carryall will deliver the harvester to the first location, even though it was cancelled well before the harvester was actually picked up. I had one case where the carryall was locked in a single facing directly above the harvester, sliding around the map while the harvester kept driving. Another case had the carryall pick up the harvester sideways.

**Comment by tovl** [comment]
The harvester sliding behind the harvester seems to be the `WaitFor` activity getting cancelled or never queued while the carryall still thinks it is locked. Somehow the activity queues get mangled but it turned out to be incredibly difficult to pinpoint exactly where things went wrong.

It is clear though that the root cause is that the carryall should not be messing with the activity queues at all, so I decided to just implement https://github.com/OpenRA/OpenRA/pull/16721#issuecomment-511197835 right now and get rid of all that.

The problem of the carryall delivering the harvester to the first location despite a new order being given seems to not be an actual regression but an old bug. This is fixed by introducing a new parent activity which can pass the destination to `DeliverUnit` at the last moment..

File: OpenRA.Mods.Common/Traits/AutoCarryall.cs [hunk 1]
```diff
@@ -96,11 +96,35 @@ void FindCarryableForTransport(Actor self)
...
+
+			protected override void OnFirstRun(Actor self)
+			{
+				QueueChild(new PickupUnit(self, cargo, 0));
+			}
+
+			public override bool Tick(Actor self)
+			{
+				var dropRange = self.Trait<Carryall>().Info.DropRange;
+				var destination = cargo.Trait<Carryable>().Destination;
```
Comment (teinarss) [inline OpenRA.Mods.Common/Traits/AutoCarryall.cs]: Wouldn't it make sense to store these in in the constructor?

Comment (tovl) [reply OpenRA.Mods.Common/Traits/AutoCarryall.cs]: It doesn't make a difference performance-wise because this will only ever tick once.

**Review by pchote — APPROVED** [review]

**Review by teinarss — APPROVED** [review]

**Comment by teinarss** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/8368130d3f4148e204301e6a2ea6b5e4379cc0e0...2e1d80a68303d4ecda744ab3bc7188476ec2e24a

**Pull request merged**